BitmapGenerator/BitmapGenerator.py

```diff
@@ -101,114 +184,214 @@ def __init__(self):
-    threeDWidget.show()
-
-    # configure the view
-    cacheViewNode = slicer.vtkMRMLViewNode()
-    cacheViewNode.Copy(viewNode)
-    viewNode.SetBoxVisible(0)
-    viewNode.SetAxisLabelsVisible(0)
-    viewNode.SetBackgroundColor((1,1,1))
-    viewNode.SetBackgroundColor2((1,1,1))
 
+    viewLayoutName = "BitmapGenerator"
+    # layout label is usually shorter than the full layout name, but in this case it helps if it is
+    # obvious that the widget belongs to this module
+    viewLayoutLabel = "BitmapGenerator"
+
+    # Retrieve or create MRML view node
+    if not self.threeDViewNode:
+      self.threeDViewNode = slicer.mrmlScene.GetSingletonNode("BitmapGenerator","vtkMRMLViewNode")
+    if not self.threeDViewNode:
+      self.threeDViewNode = slicer.mrmlScene.CreateNodeByClass("vtkMRMLViewNode")
+      self.threeDViewNode.UnRegister(None)
+      self.threeDViewNode.SetSingletonTag("BitmapGenerator")
+      self.threeDViewNode.SetName(viewLayoutName)
+      self.threeDViewNode.SetLayoutName(viewLayoutName)
+      self.threeDViewNode.SetLayoutLabel(viewLayoutLabel)
+      self.threeDViewNode.SetLayoutColor(1, 1, 0)
+      self.threeDViewNode.SetAndObserveParentLayoutNodeID(viewOwnerNode.GetID())
+      self.threeDViewNode = slicer.mrmlScene.AddNode(self.threeDViewNode)
+
+    # Create widget
+    if not self.threeDWidget:
+      self.threeDWidget = slicer.qMRMLThreeDWidget()
+      self.threeDWidget.setObjectName("ThreeDWidget"+self.threeDViewNode.GetLayoutLabel())
+      self.threeDWidget.viewLabel = self.threeDViewNode.GetLayoutLabel()
+      self.threeDWidget.viewColor = qt.QColor.fromRgbF(*self.threeDViewNode.GetLayoutColor())
+      self.threeDWidget.setMRMLScene(slicer.mrmlScene)
+      self.threeDWidget.setMRMLViewNode(self.threeDViewNode)
 
+
+  def generate(self,volumeRenderingNode,filePattern="/tmp/slice_%04d.png"): # underscore not dash
+    self.cancelRequested = False
+
+    # Ensure view node and widget exists and cross-references are up-to-date
+    self.create3dView(volumeRenderingNode)
+    self.threeDViewNode.SetAndObserveParentLayoutNodeID(volumeRenderingNode.GetID())
+    self.threeDWidget.setMRMLViewNode(self.threeDViewNode)
+
+    # Make sure the selected volume rendering node is visible in the capture view
+    if not volumeRenderingNode.IsDisplayableInView(self.threeDViewNode.GetID()):
+      volumeRenderingNode.AddViewNodeID(self.threeDViewNode.GetID())
+
+    # Configure view and widget
+    self.threeDViewNode.SetBoxVisible(0)
+    self.threeDViewNode.SetAxisLabelsVisible(0)
+    self.threeDViewNode.SetVolumeRenderingQuality(slicer.vtkMRMLViewNode.Normal) # viewNode.Maximum could provide better quality but slower
+    self.threeDViewNode.SetRenderMode(slicer.vtkMRMLViewNode.Orthographic)
+    self.threeDViewNode.SetBackgroundColor((1,1,1))
+    self.threeDViewNode.SetBackgroundColor2((1,1,1))
+
+    # Turn off shading. We do not want any lighting effect (specular reflections, etc.) to be baked into the image.
```

Comment:
This is debatable - @SlicerFab/owners have discussed wanting to have better lighting models (e.g. cinematic rendering and highlights) baked into the print, mainly for esthetic reasons even if it's not physically realistic in the setting of the print.  This is because, as of now, there's no way to control the light reflectance behavior of the printed material (other than diffuse color).

Reply:
When we get more experience with actual prints we can consider adding a checkbox for including/excluding shading.

Reply:
Baking in lighting is an interesting idea. We probably cannot rely on the volume renderer to produce this lighting because if we clip the volume using a plane then the surface normal becomes the clipping plane normal (and there are even some bugs with this, so it is not very reliable).

Using raycasting to get the colors is also somewhat unpredictable, as too thin slabs may produce noisy images and sampling artifacts; thick slabs are affected by voxels far away from the current layer. It may be better to use vtkImageReslice (with optional thick slice sampling) and use color and opacity LUTs to get color and opacity. We might also reslice the gradient image and add some simulated lighting.